Sources/LeafKit/LeafRenderer.swift

```diff
@@ -176,65 +274,87 @@ public final class LeafRenderer {
-            var lexer = LeafLexer(name: file, template: template)
-            let tokens = try lexer.lex()
-            var parser = LeafParser(name: file, tokens: tokens)
-            return try parser.parse()
+        var chain = chain
+        _ = chain.append(ast.name)
+        let intersect = ast.unresolvedRefs.intersection(Set<String>(chain))
+        guard intersect.count == 0 else {
+            let badRef = intersect.first ?? ""
+            _ = chain.append(badRef)
+            return self.eventLoop.makeFailedFuture(LeafError.cyclicalReference(badRef, chain.joined(separator: " -> ")))
         }
+
+        let fetchRequests = ast.unresolvedRefs.map { self.fetch(template: $0, chain: chain) }
         
-        return syntax.flatMap { syntax in
-            let dependencies = self.readDependencies(syntax.dependencies)
-            let resolved = dependencies.flatMapThrowing { dependencies -> ResolvedDocument in
-                let unresolved = UnresolvedDocument(name: file, raw: syntax)
-                let resolver = ExtendResolver(document: unresolved, dependencies: dependencies)
-                return try resolver.resolve(rootDirectory: self.configuration.rootDirectory)
-            }
-            
-            return resolved.flatMap { resolved in self.cache.insert(resolved, on: self.eventLoop) }
-        }
-    }
-    
-    private func readDependencies(_ dependencies: [String]) -> EventLoopFuture<[ResolvedDocument]> {
-        let fetchRequests = dependencies.map(self.fetch)
         let results = EventLoopFuture.whenAllComplete(fetchRequests, on: self.eventLoop)
-        return results.flatMapThrowing { results in
-            return try results.map { result -> ResolvedDocument in
+        return results.flatMap { results in
+            let results = results
+            var externals: [String: LeafAST] = [:]
+            for result in results {
+                // skip any unresolvable references
                 switch result {
-                case .success(let ob): return ob
-                case .failure(let e): throw e
+                    case .success(let external):
+                        guard let external = external else { continue }
+                        externals[external.name] = external
+                    case .failure(let e): return self.eventLoop.makeFailedFuture(e)
                 }
             }
+            // return new inlined AST that may potentially still have references that can't be resolved
+            let new = LeafAST(from: ast, referencing: externals)
+            return self.cache.insert(new, on: self.eventLoop, replace: true )
         }
     }
-    
+
+    private func read(name: String) -> EventLoopFuture<LeafAST?> {
+        let path = expand(template: name)
+        let raw = readBytes(file: path)
+
+        return raw.flatMapThrowing { raw -> LeafAST? in
+            var raw = raw
+
+            // MARK: Should this actually throw an error if readString can't read readableBytes?
```

Comment:
Will have to think about this one - Leaf does NOT care about UTF8 afaik, you could theoretically have a Leaf template that's handling raw binary data.

Reply:
I disagree - UTF-8 is implicit all over the place, from the use of `String` and `Character` for the parsing (consider the unit test that has an emoji as a tag indicator!) right up to the behaviors of the `ViewRenderer` in the `Leaf` provider package. In theory Leaf as a language could operate on binary, yes, but it's debatable that it would have much utility in a mode like that, and there's definitely no support for using it that way anywhere in its implementation as it stands.

Reply:
Hmm. You're right. I honestly haven't looked that much at Lex/Parse yet so I think I'd just been going off an old issue suggesting it DOESN'T inherently use UTF8 right now.